plugins/kotlin-as-java/src/main/kotlin/converters/KotlinToJavaConverter.kt

```diff
@@ -259,24 +266,56 @@ internal fun DClass.asJava(): DClass = copy(
             )
         }, // name may not always be valid here, however classNames should always be not null
     functions = functionsInJava(),
-    properties = properties
-        .filterNot { it.hasJvmSynthetic() }
-        .map { it.asJava() },
-    classlikes = classlikes.map { it.asJava() },
+    properties = propertiesInJava(),
+    classlikes = classlikesInJava(),
     generics = generics.map { it.asJava() },
+    companion = companion.asKotlinCompanion().asJava(),
     supertypes = supertypes.mapValues { it.value.map { it.asJava() } },
     modifier = if (modifier.all { (_, v) -> v is KotlinModifier.Empty }) sourceSets.associateWith { JavaModifier.Final }
     else sourceSets.associateWith { modifier.values.first() }
 )
 
+/**
+ * Companion objects requires some custom logic for rendering as Java.
+ * They are excluded from usual classlikes rendering and added after.
+ */
+internal fun DClass.classlikesInJava(): List<DClasslike> {
+    val companionAsJava = companion.asKotlinCompanion().asJava()
+    val filterOutCompanion = { it: DClasslike -> it.name != companion?.name }
+
+    val classlikes = classlikes
+        .filter { filterOutCompanion(it) }
```

Comment:
I believe it'd be better to inline `filterOutCompanion`, it doesn't big and its implementation is pretty clear